velox/exec/Task.cpp

```diff
@@ -563,178 +563,178 @@ void Task::start(
-        bufferManager->initializeTask(
-            self,
-            partitionedOutputNode->kind(),
-            partitionedOutputNode->numPartitions(),
-            totalOutputDrivers);
-      }
+  concurrentSplitGroups_ = concurrentSplitGroups;
+  // Preallocate a bunch of slots for max concurrent grouped execution
+  // drivers, if needed.
+  if (numDriversPerSplitGroup_ > 0) {
+    drivers_.resize(numDriversPerSplitGroup_ * concurrentSplitGroups_);
+  }
 
-      // NOTE: MergeExchangeNode doesn't use the exchange client created here to
-      // fetch data from the merge source but only uses it to send abortResults
-      // to the merge source of the split which is added after the task has
-      // failed. Correspondingly, MergeExchangeNode creates one exchange client
-      // for each merge source to fetch data as we can't mix the data from
-      // different sources for merging.
-      if (auto exchangeNodeId = factory->needsExchangeClient()) {
-        self->createExchangeClient(pipeline, exchangeNodeId.value());
-      }
+  // We create the drivers running pipelines in ungrouped execution mode
+  // first.
+  if (numDriversUngrouped_ > 0) {
+    // Create the drivers we are going to run for this task.
+    std::vector<std::shared_ptr<Driver>> drivers;
+    drivers.reserve(numDriversUngrouped_);
+    createSplitGroupStateLocked(kUngroupedGroupId);
+    createDriversLocked(kUngroupedGroupId, drivers);
+    if (pool_->stats().currentBytes != 0) {
+      VELOX_FAIL(
+          "Unexpected memory pool allocations during task[{}] driver initialization: {}",
+          taskId_,
+          pool_->treeMemoryUsage());
     }
 
-    std::unique_lock<std::mutex> l(self->mutex_);
-
-    // Preallocate a bunch of slots for max concurrent grouped execution
-    // drivers, if needed.
-    if (self->numDriversPerSplitGroup_ > 0) {
-      self->drivers_.resize(
-          self->numDriversPerSplitGroup_ * self->concurrentSplitGroups_);
+    // Prevent the connecting structures from being cleaned up before all
+    // split groups are finished during the grouped execution mode.
+    if (isGroupedExecution()) {
+      splitGroupStates_[kUngroupedGroupId].mixedExecutionMode = true;
     }
 
-    // We create the drivers running pipelines in ungrouped execution mode
-    // first.
-    if (self->numDriversUngrouped_ > 0) {
-      // Create the drivers we are going to run for this task.
-      std::vector<std::shared_ptr<Driver>> drivers;
-      drivers.reserve(self->numDriversUngrouped_);
-      self->createSplitGroupStateLocked(kUngroupedGroupId);
-      self->createDriversLocked(self, kUngroupedGroupId, drivers);
-      if (self->pool_->stats().currentBytes != 0) {
-        VELOX_FAIL(
-            "Unexpected memory pool allocations during task[{}] driver initialization: {}",
-            self->taskId_,
-            self->pool_->treeMemoryUsage());
+    // We might have first slots taken for grouped execution drivers, so need
```

Comment:
// Slots in the front are used by grouped execution drivers. Ungrouped execution drivers come after these.